core/port/reachability.go

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2021 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package port
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/gofrs/uuid"
+)
+
+const (
+	portFieldSize = 2
+	packetSize    = portFieldSize + uuid.Size
+	sendPackets   = 3
+)
+
+// ErrEmptyServerAddressList indicates there are no servers to get response from
+var ErrEmptyServerAddressList = errors.New("empty server address list specified")
+
+// GloballyReachable checks if UDP port is reachable from global Internet,
+// performing probe against asymmetric UDP echo server
+func GloballyReachable(ctx context.Context, port Port, echoServerAddresses []string, timeout time.Duration) (bool, error) {
+	count := len(echoServerAddresses)
+	if count == 0 {
+		return false, ErrEmptyServerAddressList
+	}
+
+	// Claim port
+	rxAddr := &net.UDPAddr{
+		Port: port.Num(),
+	}
+
+	rxSock, err := net.ListenUDP("udp", rxAddr)
+	if err != nil {
+		return false, err
+	}
+	defer rxSock.Close()
+
+	// Prepare request
+	msg := make([]byte, packetSize)
+	binary.BigEndian.PutUint16(msg, uint16(port.Num()))
+
+	probeUUID, err := uuid.NewV4()
+	if err != nil {
+		return false, err
+	}
+	copy(msg[portFieldSize:], probeUUID[:])
+
+	// Send probes. Proceed to listen after first send success.
+	sendResultChan := make(chan error)
+
+	for _, address := range echoServerAddresses {
+		go func(echoServerAddress string) {
+			err := func() error {
```

Comment:
seems like a perfect candidate to extract to a separate function

Reply:
Applied